src/transactions/TransactionUtils.cpp

```diff
@@ -1159,6 +1174,384 @@ removeOffersByAccountAndAsset(AbstractLedgerTxn& ltx, AccountID const& account,
     ltxInner.commit();
 }
 
+static bool
+tryGetEntrySponsor(LedgerTxnEntry& entry, AccountID& sponsor)
+{
+    if (entry.current().ext.v() == 1 &&
+        getLedgerEntryExtensionV1(entry.current()).sponsoringID)
+    {
+        sponsor = *getLedgerEntryExtensionV1(entry.current()).sponsoringID;
+        return true;
+    }
+
+    return false;
+}
+
+static AccountID
+getTrustLineBacker(LedgerTxnEntry& trustLine)
+{
+    AccountID sponsor;
+    if (tryGetEntrySponsor(trustLine, sponsor))
+    {
+        return sponsor;
+    }
+
+    return trustLine.current().data.trustLine().accountID;
+}
+
+static void
+prefetchForRevokeFromPoolShareTrustLines(
+    AbstractLedgerTxn& ltx, AccountID const& accountID,
+    std::vector<LedgerTxnEntry>& poolShareTrustLines)
+{
+    // first prefetch the liquidity pools and pool share trustline sponsors
+    UnorderedSet<LedgerKey> keys;
+    for (auto& trustLine : poolShareTrustLines)
+    {
+        keys.emplace(liquidityPoolKey(
+            trustLine.current().data.trustLine().asset.liquidityPoolID()));
+
+        AccountID sponsor;
+        if (tryGetEntrySponsor(trustLine, sponsor))
+        {
+            keys.emplace(accountKey(sponsor));
+        }
+    }
+    ltx.prefetch(keys);
+
+    // now prefetch the asset trustlines
+    keys.clear();
+    for (auto const& trustLine : poolShareTrustLines)
+    {
+        // prefetching shouldn't affect the protocol, so use loadWithoutRecord
+        // to not touch lastModified
+        auto pool = ltx.loadWithoutRecord(liquidityPoolKey(
+            trustLine.current().data.trustLine().asset.liquidityPoolID()));
+
+        auto const& params =
+            pool.current().data.liquidityPool().body.constantProduct().params;
+
+        if (params.assetA.type() != ASSET_TYPE_NATIVE &&
+            !isIssuer(accountID, params.assetA))
+        {
+            keys.emplace(trustlineKey(accountID, params.assetA));
+        }
+
+        if (params.assetB.type() != ASSET_TYPE_NATIVE &&
+            !isIssuer(accountID, params.assetB))
+        {
+            keys.emplace(trustlineKey(accountID, params.assetB));
+        }
+    }
+    ltx.prefetch(keys);
+}
+
+static ClaimableBalanceID
+getRevokeID(AccountID const& txSourceID, SequenceNumber txSeqNum,
+            uint32_t opIndex, PoolID const& poolID, Asset const& asset)
+{
+    HashIDPreimage hashPreimage;
+    hashPreimage.type(ENVELOPE_TYPE_POOL_REVOKE_OP_ID);
+    hashPreimage.revokeID().sourceAccount = txSourceID;
+    hashPreimage.revokeID().seqNum = txSeqNum;
+    hashPreimage.revokeID().opNum = opIndex;
+    hashPreimage.revokeID().liquidityPoolID = poolID;
+    hashPreimage.revokeID().asset = asset;
+
+    ClaimableBalanceID balanceID;
+    balanceID.type(CLAIMABLE_BALANCE_ID_TYPE_V0);
```

Comment:
I decided to ask here first before creating a new issue. Should it set the type to `CLAIMABLE_BALANCE_ID_TYPE_FROM_POOL_REVOKE` instead of `CLAIMABLE_BALANCE_ID_TYPE_V0`?

Reply:
`CLAIMABLE_BALANCE_ID_TYPE_FROM_POOL_REVOKE` was removed from the CAP. I would look over https://github.com/stellar/stellar-protocol/commits/master/core/cap-0038.md as there have been some minor changes recently.

Reply:
It still says:
> The claimable balances will have a different `ClaimableBalanceIDType` to indicate their origin.

I will open a PR in stellar-protocol repo. Thanks!

Reply:
Ah good catch

Reply:
https://github.com/stellar/stellar-protocol/pull/1048